### PR TITLE
MGMT-5400: Get physical_bytes from ghw if dmidecode was failed

### PR DESCRIPTION
**Pull request by eliorerz**
image

### CODE REVIEW

File: src/inventory/memory.go [hunk 1]
```diff
@@ -40,6 +40,32 @@ func newMemory(dependencies util.IDependencies) *memory {
 }
 
 func (m *memory) getTotalPhysicalBytes() int64 {
+
+	if physicalBytes := m.getDmidecodeTotalPhysicalBytes(); physicalBytes != 0 {
+		return physicalBytes
+	}
+
+	if physicalBytes := getGhwTotalPhysicalBytes(); physicalBytes != 0 {
```
Comment (omertuc) [inline src/inventory/memory.go:48]: `// Try to use ghw to get total physical bytes, as `dmidecode` returns 0 on some platforms`

File: src/inventory/memory.go [hunk 2]
```diff
@@ -40,6 +40,32 @@ func newMemory(dependencies util.IDependencies) *memory {
...
+		return physicalBytes
+	}
+
+	if usableBytes := m.getTotalUsabeBytes(); usableBytes != 0 {
```
Comment (omertuc) [inline src/inventory/memory.go:52]: `// Treat usable bytes as physical bytes, as `ghw` returns 0 on some platforms`

File: src/inventory/memory.go [hunk 3]
```diff
@@ -40,6 +40,32 @@ func newMemory(dependencies util.IDependencies) *memory {
 }
 
 func (m *memory) getTotalPhysicalBytes() int64 {
```
Comment (omertuc) [inline src/inventory/memory.go:42]: Return the method that was used to get the bytes
`func (m *memory) getTotalPhysicalBytes() (int64, string) {`

File: src/inventory/memory_test.go [hunk 4]
```diff
@@ -95,7 +95,8 @@ var _ = Describe("Memory test", func() {
 		dependencies.On("Execute", "dmidecode", "-t", "17").Return("", "Just an error", -1).Once()
 		dependencies.On("ReadFile", "/proc/meminfo").Return(nil, fmt.Errorf("Another error")).Once()
 		ret := GetMemory(dependencies)
-		Expect(ret).To(Equal(&models.Memory{}))
+		Expect(ret).To(Not(Equal(&models.Memory{})))
+		Expect(ret).To(Equal(&models.Memory{PhysicalBytes: getGhwTotalPhysicalBytes()}))
 	})
```
Comment (omertuc) [inline src/inventory/memory_test.go:100]: Add tests to make sure `getTotalPhysicalBytges` correctly does fallbacks

File: src/inventory/memory.go (hunk 3 above)
Comment (eliorerz) [reply src/inventory/memory.go:42]: That will be in a different PR. Need to add it to swagger first

Comment (omertuc) [reply src/inventory/memory.go:42]: No reason not to have this function return the method now & test it, but ignore it - then in the second PR just populate the swagger

File: src/inventory/memory.go [hunk 5]
```diff
@@ -40,6 +39,35 @@ func newMemory(dependencies util.IDependencies) *memory {
...
+
+	if physicalBytes := m.getDmidecodeTotalPhysicalBytes(); physicalBytes != 0 {
+		// Try to use ghw to get total physical bytes, as dmidecode returns 0 on some platforms
```
Comment (omertuc) [inline src/inventory/memory.go:44]: The comments are not placed correctly

File: src/inventory/memory_test.go [hunk 6]
```diff
@@ -92,11 +101,23 @@ var _ = Describe("Memory test", func() {
...
+		dependencies.On("ReadFile", "/proc/meminfo").Return(nil, fmt.Errorf("meminfo error")).Twice()
+		dependencies.On("Memory").Return(&mem1, nil).Once()
+
 		ret := GetMemory(dependencies)
 		Expect(ret).To(Equal(&models.Memory{}))
+		Expect(ret.PhysicalBytes).To(Equal(int64(0)))
+		Expect(ret.UsableBytes).To(Equal(int64(0)))
 	})
+	It("Physical fallback to useable", func() {
+		dependencies.On("Execute", "dmidecode", "-t", "17").Return("", "Just an error", -1).Once()
```
Comment (omertuc) [inline src/inventory/memory_test.go:114]: What about testing fallback to ghw?

File: src/inventory/memory.go (hunk 3 above)
Comment (eliorerz) [reply src/inventory/memory.go:41]: @omertuc 
Adding `physical_bytes_method` to swagger.yaml will generate ENUM that can and will be the second return value from `getTotalPhysicalBytes` on the agent.
We need to add it first and then add it in the agent. Returning any other value will cause another PR for replacing it to the new ENUM.

Comment (omertuc) [reply src/inventory/memory.go:41]: Fair enough

File: src/inventory/memory_test.go (hunk 6 above)
Comment (eliorerz) [reply src/inventory/memory_test.go:128]: @omertuc Done

File: src/inventory/memory_test.go [hunk 7]
```diff
@@ -92,11 +108,30 @@ var _ = Describe("Memory test", func() {
...
+		dependencies.On("Execute", "dmidecode", "-t", "17").Return("", "dmidecode error", -1).Once()
+		dependencies.On("ReadFile", "/proc/meminfo").Return(nil, fmt.Errorf("meminfo error")).Twice()
+		dependencies.On("Memory").Return(&mem1, nil).Once()
+
 		ret := GetMemory(dependencies)
 		Expect(ret).To(Equal(&models.Memory{}))
+		Expect(ret.PhysicalBytes).To(Equal(int64(0)))
+		Expect(ret.UsableBytes).To(Equal(int64(0)))
+	})
+	It("Physical fallback to ghw", func() {
```
Comment (omertuc) [inline src/inventory/memory_test.go:120]: "dmidecode fallback to ghw"

File: src/inventory/memory_test.go [hunk 8]
```diff
@@ -92,11 +108,30 @@ var _ = Describe("Memory test", func() {
...
+		dependencies.On("Execute", "dmidecode", "-t", "17").Return("", "Just an error", -1).Once()
+		dependencies.On("ReadFile", "/proc/meminfo").Return([]byte(meminfoContentskB), nil).Once()
+		dependencies.On("Memory").Return(&mem2, nil).Once()
+		ret := GetMemory(dependencies)
+		Expect(ret.PhysicalBytes).To(Equal(mem2.TotalPhysicalBytes))
 	})
+	It("Physical fallback to useable", func() {
```
Comment (omertuc) [inline src/inventory/memory_test.go:127]: "ghw fallback to usable"

**Pull request merged**